### DISCUSSION TITLE
Recreating UltiSnip's VISUAL

### DISCUSSION THREAD

**Discussion by Gargantuar314** [L3MON4D3/LuaSnip#1100] [Q&A] [answered]
I thought about this a long time, but could not figure it out.
This is a very specific problem.

I recreated `UltiSnip`'s `${1:${VISUAL}}` using [this post](https://www.ejmastnak.com/tutorials/vim-latex/luasnip/#advanced-nodes).
With the option `store_selection_keys = '<Tab>'`, this is the snippet file:
```lua
local function insert_wrap(_, snip) -- Emulate Ultisnips' '${1:${VISUAL}}'
    local raw = snip.env.LS_SELECT_RAW
    if #raw == 0 then -- Insert text if no selection
        return sn(nil, i(1))
    else -- Else insert selection
        return sn(nil, t(raw))
    end
end

return {}, {

s({
    trig = 'mt',
    name = 'inline math mode'
}, {
    t('$'), i(1), t('$')
}),

s({
    trig = "(",
    dscr = "parenthesis",
    wordTrig = false
}, {
    t('('), d(1, insert_wrap), t(')')
}),

s({
    trig = '_',
    name = 'subscript',
    wordTrig = false,
}, {
    t('_'),
    m(1, '^.$', '', '{'), i(1), m(1, '^.$', '', '}')
}),

}
```

The idea is to make a subscript command in Latex, but only wrap the input in braces if the input is precisely one character.
This should give `h_n`, but `h_{na}` for instance.

Now, if I type `mth_n<TAB>`, I obtain `$h_n$` as usual; the braces around the `n` disappear.
If I type `mth_n<TAB>A_k<TAB>`, I do obtain `$h_nA_k$` as expected; after the first `<TAB>`, the braces of `n` dissapear, and after the second `<TAB>` the ones of `k` disappear.

But if I type `mth_n<TAB>(A_k<TAB>`, I obtain `$h_{n(A_k)}$`; after the first `<TAB>`, the braces of `n` disappear, but after the second `<TAB>`, braces suddenly appear *including around* the `n`.
What I would expect is `$h_n(A_k)$`, and all snippets finish after tabbing out a third and fourth time (because of the parenthesis and the `$`).
Notice that everything works as expected if I **am not** in a snippet, i.e. typing `h_n<TAB>(A_k<TAB>` really does give `h_n(A_k)`.

Could someone explain to me why this is happening?
And how should I fix the function `insert_wrap` in order to remedy this?

**Comment by L3MON4D3** [comment, marked as answer]
Hey, nice catch, that's a bug!
The issue was that the extmark that marks the region of the i(1) was not adjusted correctly during expansion of the `()`-snippet, and since it had end-rgrav true, it received all the text inserted into the `()`.
I've fixed this (straightforward with your example/MWE, thank you!), and it should all work correctly on `master`.
(this bug only occurs if the new snippet is expanded inside another snippet, immediately (so no text at all between them) behind another insertNode, I guess that's why it only popped up so late :D )

**Reply by Gargantuar314** [reply]
Thanks a lot, it works!
